File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.java
import java.util.*;

class Solution {
    public int solution(int a, int b, int c, int d) {
        int answer = 0;
        Map<Integer, Integer> numMap = new HashMap<>();
        numMap.put(a, numMap.getOrDefault(a, 0) + 1);
        numMap.put(b, numMap.getOrDefault(b, 0) + 1);
        numMap.put(c, numMap.getOrDefault(c, 0) + 1);
        numMap.put(d, numMap.getOrDefault(d, 0) + 1);
        
        List<Integer> keyList = new ArrayList<>(numMap.keySet());
        if (keyList.size() == 1) {
            return 1111 * keyList.get(0);
        } else if (keyList.size() == 2) {
            int p = keyList.get(0);
            int q = keyList.get(1);
            int pValue = numMap.get(p);
            int qValue = numMap.get(q);
            
            if (pValue == qValue) {
                return (p+q)*Math.abs(p-q);
            } else {
                double calc;
                if (pValue > qValue) {
                    calc = 10 * p + q;
                } else {
                    calc = 10 * q + p;
                }
                return (int)Math.pow(calc, 2.0);
            }
        } else if (keyList.size() == 3) {
            int p = keyList.get(0); int pValue = numMap.get(p);
            int q = keyList.get(1); int qValue = numMap.get(q);
            int r = keyList.get(2); int rValue = numMap.get(r);
            
            if (pValue == qValue) {
                return p * q;
            } else if (pValue == rValue){
                return p * r;
            } else {
                return q * r;
            }
        } else {
            Collections.sort(keyList);
            return keyList.get(0);
        }
    }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.kt
/**
 * 해쉬맵에 저장해서 개수로 판별
 * 1인 경우: 모든 같은 숫자
 * 2인 경우: 3, 1로 나눠지면 2번 케이스, 아니면 3번 케이스
 * 3인 경우: p, q, r
 * 4인 경우: 제일 작은 숫자
 */
import kotlin.collections.*
import kotlin.math.*

class Solution {
    fun solution(a: Int, b: Int, c: Int, d: Int): Int {
        val numMap = hashMapOf<Int, Int>()
            .apply {
                put(a, getOrDefault(a, 0) + 1)
                put(b, getOrDefault(b, 0) + 1)
                put(c, getOrDefault(c, 0) + 1)
                put(d, getOrDefault(d, 0) + 1)
            }

        
        return when (numMap.size) {
            1 -> 1111 * numMap.keys.first()
            2 -> {
                val (p, q) = numMap.keys.sortedByDescending { numMap[it] }
                if (numMap[p] == numMap[q]) {
                    (p+q) * abs(p-q)
                } else {
                    (10 * p + q.toFloat()).pow(2).toInt()
                }
            }
            3 -> {
                val (p, q, r) = numMap.keys.sortedByDescending { numMap[it] }
                q * r
            }
            4 -> numMap.keys.minOf { it }
            else -> 0
        }
    }
}